src/euphorie/client/model.py

```diff
@@ -1031,6 +1038,38 @@ def tool(self):
         client = api.portal.get().client
         return client.restrictedTraverse(str(self.zodb_path), None)
 
+    def update_completion_percentage(self):
+        query = (
+            Session.query(SurveyTreeItem)
+            .filter(SurveyTreeItem.session_id == self.id)
+            .filter(SurveyTreeItem.type == "module")
+            .filter(SurveyTreeItem.skip_children == False)
+        )
+        total_risks = 0
+        answered_risks = 0
+        for module in query:
+            if not module.path:
+                # XXX When does a module not have a path?
+                continue
+            total_risks_query = (
+                Session.query(Risk)
+                .filter(Risk.session_id == self.id)
+                .filter(Risk.type == "risk")
+                .filter(Risk.path.like(module.path + "%"))
+                .filter(Risk.depth == module.depth + 1)
+            )
+            total_risks = total_risks + total_risks_query.count()
+            answered_risks_query = (
+                total_risks_query
+                .filter(Risk.identification != None)
+            )
+            answered_risks = answered_risks + answered_risks_query.count()
+
+        self.completion_percentage = (
+            int(round((answered_risks / total_risks) * 100.))
+            if total_risks else 0.
+        )
+
```

Comment:
I think that this approach will not be able to deal with all complexities in OiRA. Example: If a an optional module is turned off, and this module has submodules, will the risks inside the submodules be correctly discarded from the equation? (See below my comment on the test)

And it seems to me that we need an awful number of queries. Let me see if we might have examples that combine some of the steps into a single query.